src/main/java/io/cdap/plugin/gcp/bigtable/common/BigtableErrorDetailsProvider.java

```diff
@@ -0,0 +1,115 @@
+import com.google.common.base.Throwables;
+import io.cdap.cdap.api.exception.ErrorCategory;
+import io.cdap.cdap.api.exception.ErrorCodeType;
+import io.cdap.cdap.api.exception.ErrorType;
+import io.cdap.cdap.api.exception.ErrorUtils;
+import io.cdap.cdap.api.exception.ProgramFailureException;
+import io.cdap.cdap.etl.api.exception.ErrorContext;
+import io.cdap.plugin.gcp.common.GCPErrorDetailsProvider;
+import io.cdap.plugin.gcp.common.GCPUtils;
+
+import org.apache.hadoop.hbase.client.RetriesExhaustedWithDetailsException;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * A custom ErrorDetailsProvider for BigTable plugins.
+ */
+public class BigtableErrorDetailsProvider extends GCPErrorDetailsProvider {
+  private static final String ERROR_MESSAGE_FORMAT = "Error occurred in the phase: '%s'. Error message: %s";
+
+  static Map<Status.Code, Integer> actionErrorMap = new HashMap<>();
+
+  static {
+    actionErrorMap.put(Status.Code.CANCELLED, 499);
+    actionErrorMap.put(Status.Code.UNKNOWN, 500);
+    actionErrorMap.put(Status.Code.INVALID_ARGUMENT, 400);
+    actionErrorMap.put(Status.Code.DEADLINE_EXCEEDED, 504);
+    actionErrorMap.put(Status.Code.NOT_FOUND, 404);
+    actionErrorMap.put(Status.Code.ALREADY_EXISTS, 409);
+    actionErrorMap.put(Status.Code.PERMISSION_DENIED, 403);
+    actionErrorMap.put(Status.Code.UNAUTHENTICATED, 401);
+    actionErrorMap.put(Status.Code.RESOURCE_EXHAUSTED, 429);
+    actionErrorMap.put(Status.Code.FAILED_PRECONDITION, 400);
+    actionErrorMap.put(Status.Code.ABORTED, 409);
+    actionErrorMap.put(Status.Code.OUT_OF_RANGE, 400);
+    actionErrorMap.put(Status.Code.UNIMPLEMENTED, 501);
+    actionErrorMap.put(Status.Code.INTERNAL, 500);
+    actionErrorMap.put(Status.Code.UNAVAILABLE, 503);
+    actionErrorMap.put(Status.Code.DATA_LOSS, 500);
+  }
+
+  @Override
+  protected String getExternalDocumentationLink() {
+    return GCPUtils.BIG_TABLE_SUPPORTED_DOC_URL;
+  }
+
+  @Override
+  public ProgramFailureException getExceptionDetails(Exception e, ErrorContext errorContext) {
+    ProgramFailureException ex = super.getExceptionDetails(e, errorContext);
+    if (ex != null) {
+      return ex;
+    }
+    List<Throwable> causalChain = Throwables.getCausalChain(e);
+    for (Throwable t : causalChain) {
+      if (t instanceof StatusRuntimeException) {
+        return getProgramFailureExceptionFromBigTableException((StatusRuntimeException) t);
+      }
+      // Some RPC exception may be wrapped in a RetriesExhaustedWithDetailsException
+      if (t instanceof RetriesExhaustedWithDetailsException) {
+        RetriesExhaustedWithDetailsException r = (RetriesExhaustedWithDetailsException) t;
+        List<Throwable> innerCauses = r.getCauses();
+        for (Throwable innerCause : innerCauses) {
+          if (innerCause instanceof StatusRuntimeException) {
+            return getProgramFailureExceptionFromBigTableException((StatusRuntimeException) innerCause);
```

Comment:
for every `innerCause` we can just call `this.getExceptionDetails(innerCause, errorContext)`, if it is not null, we can continue in the loop otherwise return the exception?